README.md
# rPELMO

The R package **rPELMO** provides some utilities for working with the pesticide
leaching model FOCUS PELMO.

## Installation

The easiest way to install the package is probably to use 
[drat](https://cran.r-project.org/package=drat):

```r
install.packages("drat")
drat::addRepo("jranke")
install.packages("rPELMO")
```

Alternatively you can install the package 
using the `devtools` package.  Using `quick = TRUE` skips docs,
multiple-architecture builds, demos, and vignettes.


```r
library(devtools)
install_github("jranke/rPELMO", quick = TRUE)
```

## Use

Please refer to the [reference](http://pkgdown.jrwb.de/rPELMO/reference/index.html).

## Examples

An example how FOCUS PELMO can be run in parallel under Linux is shown
[here](http://pkgdown.jrwb.de/rPELMO/reference/PELMO_runs.html).


ChangeLog
commit 4fec115b6839a0ef1c05278a840ce0b18e5be3e9
Author: Johannes Ranke <jranke@uni-bremen.de>
Date:   2019-01-31 01:24:39 +0100

    Fix a link

commit 1d59a7e731e2dd0721825c8c965a30d9c28a4789
Author: Johannes Ranke <jranke@uni-bremen.de>
Date:   2019-01-31 01:23:08 +0100

    Trivial commit to initialise pkgdown

commit e6b7fd5af92bedd1619839dbf476a5b6c43582dc
Author: Johannes Ranke <jranke@uni-bremen.de>
Date:   2019-01-31 00:02:34 +0100

    Initial commit
    
    The functions for running PELMO were separated out from the pfm package.
    The function PELMO_psm for creating FOCUS PELMO input files is published
    for the first time.


tests/testthat.R
library(testthat)
library(pfm)

test_check("pfm")


tests/testthat/test_PELMO.R
library(testthat)
library(rPELMO)
context("Create PELMO runs from psm files and execute them")
PELMO_base <- system.file("FOCUSPELMO.553", package = "PELMO.installeR")

test_archive <- system.file("testdata/FOCUS_PELMO.tar.bz2", package = "rPELMO")
test_dir <- tempdir()
untar(test_archive, exdir = test_dir, compressed = "bzip2")

runs <- list(
  list(
    psm = "Pesticide_D",
    fbe = c("Por"),
    vbe = c("Por")),
  list(
    psm = "Pesticide_D_1_day_pre_em_every_third_year",
    pot = c("Cha", "Ham")),
  list(
    psm = "Pesticide_D_1_May_every_other_year_mets",
    win = names(pfm::FOCUS_GW_scenarios_2012$names)))

# Check if we have wine on the path
wine_installed <- system('wine --version', ignore.stdout = TRUE) == 0
skip_if_no_wine <- function() {
  if (!wine_installed) skip("wine is not available")
}

test_that("PELMO paths are correctly created", {
  skip_if_no_wine()
  psm_paths = c(
     PELMO_path(runs[[1]]$psm, "fbe", "Por"),
     PELMO_path(runs[[2]]$psm, "pot", "Ham"),
     PELMO_path(runs[[3]]$psm, "win", "Cha"))

  for (i in seq_along(psm_paths)) {
    psm_file <- file.path(test_dir, psm_paths[i], paste0(runs[[i]]$psm, ".psm"))
    expect_true(file.exists(psm_file))
    psm_new_location <- file.path(PELMO_base, basename(psm_file))
    file.copy(psm_file, psm_new_location, overwrite = TRUE)
  }
})

test_that("PELMO runs are correctly set up", {
  skip_if_no_wine()

  # Prepare runs in analogy to the test archive
  skip_on_cran()
  PELMO_runs(runs, psm_dir = PELMO_base, execute = FALSE, evaluate = FALSE, overwrite = TRUE)

  # Check that input files are correctly generated in the right location
  for (run in runs) {
    psm <- run$psm
    crops <- setdiff(names(run), "psm")
    for (crop in crops) {
      for (scenario in run[[crop]]) {
        pp <- PELMO_path(psm, crop, scenario)

        input_new <- readLines(file.path(PELMO_base, "FOCUS", pp, "pelmo.inp"))
        input_test <- readLines(file.path(test_dir, pp, "pelmo.inp"))

        # Check if the input files are correctly reproduced
        expect_identical(input_new, input_test)
      }
    }
  }
})

test_that("PELMO runs can be run and give the expected result files", {
  skip_if_no_wine()
  skip_on_cran()

  run_PELMO(runs, cores = 15)

  plm_files <- c("CHEM.PLM", "ECHO.PLM",
                 "KONZCHEM.PLM", "KONZC_A1", "KONZC_B1",
                 "PLNTPEST.plm", "PLOT.PLM", "WASSER.PLM")

  # Check that if output is the same as in the test archive
  for (run in runs) {
    psm <- run$psm
    crops <- setdiff(names(run), "psm")
    for (crop in crops) {
      for (scenario in run[[crop]]) {
        pp <- PELMO_path(psm, crop, scenario)

        for (plm in plm_files) {
          if (file.exists(file.path(test_dir, pp, plm))) {
            new <- readLines(file.path(PELMO_base, "FOCUS", pp, plm))
            test <- readLines(file.path(test_dir, pp, plm))

            # Don't check for differences in the PESTICIDE BALANCE ERROR
            pest_balance_error <- suppressWarnings(grep("PESTICIDE BALANCE ERROR", new))
            # Suppress warnings about invalid strings in this locale caused by the files

            # Check if the ouput files are correctly reproduced
            expect_identical(new[!pest_balance_error], test[!pest_balance_error])
          }
        }
      }
    }
  }
})

test_that("PELMO runs are correctly evaluated", {
  skip_if_no_wine()
  skip_on_cran()

  pfm_PECgw <- evaluate_PELMO(runs)

  # Check if output is the same as in the test archive
  for (run in runs) {
    psm <- run$psm
    crops <- setdiff(names(run), "psm")
    for (crop in crops) {
      for (scenario in run[[crop]]) {
        pp <- PELMO_path(psm, crop, scenario)

        period_file <- readLines(file.path(test_dir, pp, "period.plm"), encoding = "latin1")

        result_lines <- grep("^\tResults for.*in the percolate at 1 m soil depth$", period_file)
        acronyms <- gsub(".*\\((.*)\\).*", "\\1", period_file[result_lines])
        names(result_lines) <- acronyms

        results <- list()
        for (acronym in acronyms) {
          results[[acronym]] <- list()
          conc_lines <- result_lines[acronym] + 5:24
          tmp <- read.table(text = period_file[conc_lines], sep = "\t")
          results[[acronym]]$periods <- data.frame(
            period = as.integer(tmp$V2),
            flux = tmp$V3,
            percolate = tmp$V4,
            conc = tmp$V5)
          tmp80 <- read.table(text = period_file[result_lines[acronym] + 27], sep = "\t")
          results[[acronym]]$focus <- tmp80[[1, "V5"]]
        }

        period_pfm_file <- file.path(PELMO_base, "FOCUS", pp, "period_rPELMO.rda")
        load(period_pfm_file)

        #message(psm, " ", crop, " ", scenario)

        # Test for equality of all the components separately,
        # as we need to adapt the tolerance
        for (acronym in acronyms) {
          p_pelmo <- results[[acronym]]$periods
          p_test <- results[[acronym]]$periods
          expect_equal(p_test$flux, p_pelmo$flux, tol = 1e-6)
          expect_equal(p_test$percolate, p_pelmo$percolate)
          # PELMO sets the concentration to 0 when the percolate is zero.
          # We get NaN, which is more reasonable, but we need to
          # take this into account for testing
          p_test$conc <- ifelse(is.na(p_test$conc), 0, p_test$conc)
          expect_equal(p_test$conc, p_pelmo$conc, tol = 1e-3, scale = 1)

          # FOCUS PEC
          expect_equal(results[[acronym]]$focus, results[[acronym]]$focus,
                       tol = 1e-3, scale = 1)
        }
      }
    }
  }
})

test_that("PECgw from FOCUS summary files can be reproduced", {
  skip_if_no_wine()
  skip_on_cran()

  pfm_PECgw <- evaluate_PELMO(runs)

  focus_summary <- list()

  for (run in runs) {
    psm <- run$psm
    focus_summary[[psm]] <- list()

    crops <- setdiff(names(run), "psm")
    for (crop in crops) {
      scenarios <- run[[crop]]

      # Read contents of summary text file copied from the GUI output. We only
      # have results for one crop per psm file, so the crop is not in the file
      # name.
      sumfile_path <- system.file(paste0("testdata/FOCUS_Summary_", psm,
                                         ".txt"), package = "rPELMO")
      sumfile <- readLines(sumfile_path, encoding = "latin1")
      result_anchors <- grep("Results for", sumfile)
      acronyms <- gsub(".*\\((.*)\\).*", "\\1", sumfile[result_anchors])
      names(result_anchors) <- acronyms
      focus_summary[[psm]][[crop]] <- matrix(nrow = length(scenarios), ncol = length(acronyms),
                                             dimnames = list(scenarios, acronyms))
      for (acronym in acronyms) {
        tmp <- sumfile[result_anchors[acronym] + 4 + (1:length(scenarios))]
        tmp_frame <- read.table(text = tmp, sep = "\t")
        PECgw <- tmp_frame$V5
        focus_summary[[psm]][[crop]][, acronym] <- PECgw
      }
    }
  }
  expect_equal(pfm_PECgw, focus_summary)
})


R/FOCUS_PELMO_data.R
# Copyright (C) 2017-2019  Johannes Ranke
# Contact: jranke@uni-bremen.de
# This file is part of the R package rPELMO
# Until January 2019 it was part of the R package pfm

# This program is free software: you can redistribute it and/or modify it under
# the terms of the GNU General Public License as published by the Free Software
# Foundation, either version 3 of the License, or (at your option) any later
# version.

# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
# details.

# You should have received a copy of the GNU General Public License along with
# this program. If not, see <http://www.gnu.org/licenses/>

#' FOCUS PELMO crop names
#'
#' A named character vector with the crop names used in the PELMO 5.5.3 GUI.
#' For the names, three letter codes were constructed by generally taking the
#' first three letters in lower case. Only when there is an expression in
#' parentheses, the first letter of this expression becomes the first letter
#' in the three letter code, i.e. 'Peas (animals)' has the code \code{ape}.
#'
#' @docType data
#' @export
#' @examples
#' print(FOCUS_PELMO_crops)
FOCUS_PELMO_crops = c(
  app = "Apples",
  gra = "Grass and alfalfa",
  pot = "Potatoes",
  sug = "Sugar beets",
  win = "Winter cereals",
  fbe = "Beans (field)",
  vbe = "Beans (vegetables)",
  bus = "Bushberries",
  cab = "Cabbage",
  car = "Carrots",
  cit = "Citrus",
  cot = "Cotton",
  lin = "Linseed",
  mai = "Maize",
  soi = "Oil seed rape (summer)",
  woi = "Oil seed rape (winter)",
  oni = "Onions",
  ape = "Peas (animals)",
  soy = "Soybeans",
  spr = "Spring cereals",
  str = "Strawberries",
  sun = "Sunflower",
  tob = "Tobacco",
  tom = "Tomatoes",
  vin = "Vines")

#' FOCUS PELMO crop acronyms used for naming sceneario files
#'
#' A named character vector with the crop acronyms used in FOCUS PELMO 5.5.3
#' for naming the .sze files located in the FOCUS directory. The crop acronyms
#' in the file names are sometimes capitalized, sometimes not. The scenario
#' files used for Beans (field) and Beans (vegetable) are the same.
#'
#' @docType data
#' @export
#' @examples
#' print(FOCUS_PELMO_crop_sze_names)
FOCUS_PELMO_crop_sze_names = c(
  app = "apples",
  gra = "grass",
  pot = "potato",
  sug = "sbeets",
  win = "wcerea",
  fbe = "beans",
  vbe = "beans", # Same sze as for fbe is used, with 'irrigation' which has no effect on the run
  bus = "bushb",
  cab = "cabbag",
  car = "carrot",
  cit = "citrus",
  cot = "cotton",
  lin = "linse",
  mai = "maize",
  soi = "rapesu",
  woi = "rapewi",
  oni = "onions",
  ape = "peas",
  soy = "soyb",
  spr = "scerea",
  str = "strawb",
  sun = "sunflo",
  tob = "tobacc",
  tom = "tomato",
  vin = "vines")

#' Location codes in FOCUS PELMO
#'
#' A named character vector of one letter location codes used in FOCUS PELMO,
#' indexed by three letter acronyms.

#' @docType data
#' @export
#' @examples
#' print(FOCUS_PELMO_location_codes)
FOCUS_PELMO_location_codes = c(
  Cha = "C", Ham = "H", Jok = "J",
  Kre = "K", Oke = "N", Pia = "P",
  Por = "O", Sev = "S", Thi = "T")


R/PELMO_runs.R
# Copyright (C) 2017-2019  Johannes Ranke
# Contact: jranke@uni-bremen.de
# This file is part of the R package rPELMO
# Until January 2019 it was part of the R package pfm

# This program is free software: you can redistribute it and/or modify it under
# the terms of the GNU General Public License as published by the Free Software
# Foundation, either version 3 of the License, or (at your option) any later
# version.

# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
# details.

# You should have received a copy of the GNU General Public License along with
# this program. If not, see <http://www.gnu.org/licenses/>

#' Set up, execute and evaluate FOCUS PELMO runs
#'
#' Per default, the runs are not only set up but also executed with FOCUS
#' PELMO, the results are processed and returned. Currently, only FOCUS PELMO
#' as installed on Linux (or other Unix systems)
#' using the \code{install_PELMO} from the \code{PELMO.installeR} package
#' maintained on github is supported. In such installations, FOCUS PELMO is
#' installed into the package installation directory of \code{PELMO.installeR}
#' and run using \code{wine}.
#'
#' As a side effect, an R data file (period_rPELMO.rda) is generated in each
#' run directory, holding the results for all FOCUS periods, equivalent to
#' the period.plm file generated by the FOCUS PELMO GUI.
#'
#' @param runs A list of lists. Each inner lists has an element named 'psm'
#'   that holds the psm string, and elements named using three letter crop acronyms,
#'   as used in \code{\link{FOCUS_PELMO_crops}},
#'   that hold character vectors of three letter scenario acronyms
#'   as used in \code{\link{FOCUS_GW_scenarios_2012}}.
#' @param psm_dir The directory where the psm files are located
#' @param version The FOCUS PELMO version
#' @param PELMO_base Where the FOCUS PELMO installation is located
#' @param execute Should PELMO be executed directly?
#' @param cores The number of cores to execute PELMO runs in parallel
#' @param evaluate Should the results be returned?
#' @param overwrite Should existing run directories be overwritten?
#' @importFrom utils read.table
#' @importFrom pfm FOCUS_GW_scenarios_2012
#' @return If evaluate is TRUE, a list of lists of matrices holding the
#'   PEC data.
#' @references PELMO.installeR \url{https://pkgdown.jrwb.de/PELMO.installeR}
#'
#' Wine \url{https://winehq.org}
#'
#' PELMO test results \url{http://esdac.jrc.ec.europa.eu/public_path/projects_data/focus/gw/models/pelmo/test-results/test_results_FOCUS_PELMO_5_5_3.doc}
#' @export
#' @examples
#' # Reproduce the official test results for annual application of Pesticide D
#' # to winter cereals at the day before emergence
#' runs_1 <- list(
#'   list(psm = 'Pesticide_D',
#'        win = c("Cha", "Ham", "Jok", "Kre", "Oke", "Pia", "Por", "Sev", "Thi")),
#'   list(psm = 'Pesticide_D_1_day_pre_em_every_third_year',
#'        pot = c("Cha", "Ham", "Jok", "Kre", "Oke", "Pia", "Por", "Sev", "Thi")))
#' time_1 <- system.time(
#'   PECgw_1 <- PELMO_runs(runs_1, psm_dir = system.file("testdata", package = "rPELMO"),
#'     cores = 15, overwrite = TRUE)
#' )
#' print(PECgw_1)
#' # We get exactly the same PECgw values (on Linux, calling PELMO using Wine).
#' print(time_1)
#' if(!inherits(try(cpuinfo <- readLines("/proc/cpuinfo")), "try-error")) {
#'   cat(gsub("model name\t: ", "CPU model: ", cpuinfo[grep("model name", cpuinfo)[1]]))
#' }
#'
#' # Demonstrate some results with metabolites.
#' runs_2 <- list(list(psm = 'Pesticide_D_1_May_every_other_year_mets',
#'                     win = c("Cha", "Ham", "Kre")))
#' PECgw_2 <- PELMO_runs(runs_2, psm_dir = system.file("testdata", package = "rPELMO"),
#'   cores = 3, overwrite = TRUE)
#' print(PECgw_2)
PELMO_runs <- function(runs, psm_dir = ".", version = "5.5.3", PELMO_base = "auto",
                       execute = TRUE, cores = getOption("mc.cores", 2L),
                       evaluate = TRUE, overwrite = FALSE)
{
  if (PELMO_base[1] == "auto") {
    PELMO_base = file.path(system.file(package = "PELMO.installeR"),
                                       paste0("FOCUSPELMO.", gsub("\\.", "", version)))
  }
  PELMO_exe = list.files(PELMO_base, "^pelmo.*.exe", full.names = TRUE)
  if (length(PELMO_exe) == 0) {
    stop("Could not find PELMO executable. Did you run PELMO.installeR::install_PELMO()?")
  }

  run_list <- create_run_list(runs, psm_dir = psm_dir, check_psm_files = TRUE)

  setup_run <- function(x) {
    psm <- x$psm
    crop <- x$crop
    scenario <- x$scenario

    location_code <- FOCUS_PELMO_location_codes[scenario]

    run_dir <- file.path(PELMO_base, "FOCUS", PELMO_path(psm, crop, scenario))

    if (dir.exists(run_dir)) {
      if (overwrite) {
        unlink(run_dir, recursive = TRUE)
      } else {
        stop("Run directory for ", crop, " in ", scenario, "\n", run_dir, "\nalready exists")
      }
    }

    # Create the run directory
    dir.create(run_dir, recursive = TRUE)

    # Copy the psm file
    psm_file <- file.path(psm_dir, paste0(psm, ".psm"))
    file.copy(psm_file, run_dir)

    # Copy Haude factors
    file.copy(file.path(PELMO_base, "FOCUS", "HAUDE.DAT"), file.path(run_dir, "Haude.dat"))

    # Copy scenario file
    sze_file_upper <- paste0(location_code, "_",
                             toupper(FOCUS_PELMO_crop_sze_names[crop]), ".sze")
    sze_file_lower <- paste0(location_code, "_",
                             FOCUS_PELMO_crop_sze_names[crop], ".sze")
    if (file.exists(file.path(PELMO_base, "FOCUS", sze_file_lower))) {
      sze_file <- sze_file_lower
    } else {
      if (file.exists(file.path(PELMO_base, "FOCUS", sze_file_upper))) {
        sze_file <- sze_file_upper
      } else {
        stop("Could not find szenario file for ", crop, " in ", scenario)
      }
    }
    file.copy(file.path(PELMO_base, "FOCUS", sze_file), file.path(run_dir, sze_file_lower))

    # Generate PELMO input file
    input_file <- file(file.path(run_dir, "pelmo.inp"), encoding = "latin1", open = "w+")
    on.exit(close(input_file))

    add <- function(x) cat(paste0(x, "\r\n"), file = input_file, append = TRUE)

    interval <- get_interval(psm_file, location_code)

    n_years <- switch(as.character(interval),
      "1" = 26,
      "2" = 46,
      "3" = 66)

    # First line with number of climate years
    add(paste0(formatC(n_years, width = 3, flag = "0"),
               " 01 01 31 12 Version ", substr(version, 1, 1)))

    # Second line with scenario file (why did we copy it, if this line refers to the FOCUS dir...
    add(paste0("..\\..\\..\\", sze_file_lower))

    # Third line with psm file
    add(basename(psm_file))

    # Lines with climate files
    for (year in 1:n_years) {
      add(paste0("..\\..\\..\\", location_code, "_", formatC(year, width = 2, flag = "0"), ".cli"))
    }

    # Output control section
    add("00000015 00.")
    add("    PRSN    TSER  000.0     1.0 00000001")
    add("    TETD    TSER  000.0     1.0 00000001")
    add("    INFL    TSER  100.0     1.0 00000001")
    add("    RUNF    TSER  000.0     1.0 00000001")
    add("    THET    TSER  000.0     1.0 00000001")
    add("    THET    TSER  030.0     1.0 00000001")
    add("    TEMP    TSER  000.0     1.0 00000001")
    add("    TEMP    TSER  030.0     1.0 00000001")
    add("    TPAP    TSER  000.0  1.0E05 00000001")
    add("    TDKF    TSER  000.0  1.0E05 00000001")
    add("    TUPF    TSER  000.0  1.0E05 00000001")
    add("    TPST    TSER  005.0  1.0E06 00000001")
    add("    PFLX    TSER  100.0  1.0E05 00000001")
    add("    RFLX    TSER  000.0  1.0E05 00000001")
    add("    LEAC    TSER  100.0  1.0E09 00000001")

    # Copy pelmo executable
    file.copy(PELMO_exe, run_dir)

    # In addition to the files copied by the FOCUS PELMO GUI, we
    # need the error file in the run directory, as we start
    # the exe file from this directory
    file.copy(file.path(PELMO_base, "lf90.eer"), run_dir)
  }

  lapply(run_list, setup_run)

  if (execute) {
    run_PELMO(runs, version = version, PELMO_base = PELMO_base, cores = cores)
  }

  if (evaluate) {
    rPELMO_PECgw <- evaluate_PELMO(runs, version = version, PELMO_base = PELMO_base)
    return(rPELMO_PECgw)
  } else {
    invisible(NULL)
  }
}

#' @inheritParams PELMO_runs
#' @importFrom parallel mclapply
#' @rdname PELMO_runs
#' @export
run_PELMO <- function(runs, version = "5.5.3", PELMO_base = "auto",
                      cores = getOption("mc.cores", 2L))
{

  if (PELMO_base[1] == "auto") {
    PELMO_base = file.path(system.file(package = "PELMO.installeR"),
                                       paste0("FOCUSPELMO.", gsub("\\.", "", version)))
  }

  pelmo_exe = list.files(PELMO_base, "^pelmo.*.exe")

  # Create list of runs to traverse using mclapply
  run_list <- create_run_list(runs, check_psm_files = FALSE)

  # In order to run in parallel, we need to make a directory tree starting
  # from the base of the PELMO installation, as pelmo401.exe seems to look
  # for ..\..\..\summary.PLM and does not start if this is in use by another
  # instance
  run_exe <- function(x) {
    psm <- x$psm
    crop <- x$crop
    scenario <- x$scenario
    run_dir <- file.path(PELMO_base, "FOCUS", PELMO_path(psm, crop, scenario))
    exe_dir <- file.path(PELMO_base, paste(sample(c(0:9, letters, LETTERS),
                                                  size = 8, replace = TRUE),
                                           collapse = ""))
    dir.create(exe_dir)

    # Fresh PELMO base directory for every run with random name
    run_dir_exe <- file.path(exe_dir, "FOCUS", PELMO_path(psm, crop, scenario))
    dir.create(run_dir_exe, recursive = TRUE)

    # Copy the contents of the run directory
    run_files <- list.files(run_dir, full.names = TRUE)
    file.copy(run_files, run_dir_exe)

    # Copy FOCUS files
    location_code <- FOCUS_PELMO_location_codes[x$scenario]
    location_files <- list.files(file.path(PELMO_base, "FOCUS"), paste0("^", location_code),
                                 full.names = TRUE)
    focus_dir_exe <- file.path(exe_dir, "FOCUS")
    file.copy(location_files, focus_dir_exe)
    file.copy(file.path(PELMO_base, "FOCUS", "FOCUSCHECK.DAT"), focus_dir_exe)

    # We need to go the directory to simplify calling pelmo with wine
    setwd(run_dir_exe)
    psm_file <- paste0(psm, ".psm")
    message("Starting ", pelmo_exe, " with ", psm_file)
    system(paste("wine", pelmo_exe, psm_file), ignore.stdout = TRUE)

    # Copy the results to the original run directory
    plm_files <- list.files(run_dir_exe, ".plm$", full.names = TRUE)
    PLM_files <- list.files(run_dir_exe, ".PLM$", full.names = TRUE)
    file.copy(c(plm_files, PLM_files), run_dir)

    # Clean up
    unlink(exe_dir, recursive = TRUE)
  }

  mclapply(run_list, run_exe, mc.cores = cores)
}

#' Create a path of run directories as the PELMO GUI does
#'
#' @importFrom utils data
#' @param psm The psm identifier
#' @param crop The PELMO crop acronym
#' @param scenario The scenario
#' @export
PELMO_path <- function(psm, crop, scenario = NA) {
  if (crop %in% names(FOCUS_PELMO_crops)) {
    crop <- FOCUS_PELMO_crops[crop]
  }
  if (!crop %in% FOCUS_PELMO_crops) stop(crop, " is not in FOCUS_PELMO_crops.")
  psm_dir <- paste0(psm, ".run")
  crop_dir <- paste0(gsub(" ", "_-_", crop), ".run")

  # Deal with 'irrigation'. It only affects the naming of the scenario directory,
  # but it does not appear to affect the PELMO run. This can be seen when
  # comparing runs set up for Beans (field) and Beans (vegetable) for the Porto
  # scenario
  irrigation_string <- ""

  if (is.na(scenario)) {
    return(file.path(psm_dir, crop_dir))
  } else {
    # 'Irrigation' is only possible in the GUI for five scenarios
    if (scenario %in% c("Cha", "Pia", "Por", "Sev", "Thi")) {
      crop_acronyms <- names(FOCUS_PELMO_crops)
      names(crop_acronyms) <- FOCUS_PELMO_crops
      crop_acronym <- crop_acronyms[[crop]]
      # Some crops are not 'irrigated' according to the GUI
      if (crop_acronym %in% c("win", "fbe", "woi", "ape", "spr")) {
        irrigation_string <- "--_-_no_-_irrigation"
      } else {
        irrigation_string <- "--_-_irrigated"
      }
    }

    scenario_dir <- paste0(
      FOCUS_GW_scenarios_2012$names[scenario], "_-_(", FOCUS_PELMO_location_codes[scenario], ")",
      irrigation_string, ".run")

    return(file.path(psm_dir, crop_dir, scenario_dir))
  }
}

#' Create a list of runs that we can traverse
#'
#' @inheritParams PELMO_runs
#' @param check_psm_files Should we check if the psm file exists
create_run_list <- function(runs, psm_dir = ".", check_psm_files = FALSE) {
  i <- 0
  run_list <- list()
  for (run in runs) {
    psm <- run$psm
    if (check_psm_files) {
      psm_file <- file.path(psm_dir, paste0(psm, ".psm"))
      if (file.access(psm_file) == -1) {
        stop(psm_file, " is not readable")
      }
    }
    crops <- setdiff(names(run), "psm")
    for (crop in crops) {
      crop_acronyms <- names(FOCUS_PELMO_crops)
      names(crop_acronyms) <- FOCUS_PELMO_crops
      if (!crop %in% crop_acronyms) {
        if (crop %in% FOCUS_PELMO_crops) {
          crop <- crop_acronyms[crop]
        } else {
          stop("Invalid crop specification ", crop)
        }
      }
      for (scenario in run[[crop]]) {
        i <- i + 1
        run_list[[i]] <- list(psm = psm, crop = crop, scenario = scenario)
      }
    }
  }
  return(run_list)
}

#' @rdname PELMO_runs
#' @inheritParams PELMO_runs
#' @importFrom parallel mclapply
#' @export
evaluate_PELMO <- function(runs, version = "5.5.3", PELMO_base = "auto")
{

  if (PELMO_base[1] == "auto") {
    PELMO_base = file.path(system.file(package = "PELMO.installeR"),
                                       paste0("FOCUSPELMO.", gsub("\\.", "", version)))
  }

  rPELMO_PECgw <- list()
  for (run in runs) {
    psm <- run$psm
    rPELMO_PECgw[[psm]] <- list()
    crops <- setdiff(names(run), "psm")

    # Get acronyms of simulated compounds
    example_run_dir <- file.path(PELMO_base, "FOCUS", PELMO_path(psm, crops[1], run[[crops[1]]][1]))
    example_echo_file <- readLines(file.path(example_run_dir, "ECHO.PLM"), encoding = "latin1")
    parm_lines <- grep("\\*\\*\\* PARAMETERS OF", example_echo_file, value = TRUE)
    acronyms <- gsub(".*\\((.*)\\).*", "\\1", parm_lines)
    met_codes <- gsub(".*METABOLITE (..).*", "\\1", parm_lines)
    met_codes[1] <- NA
    names(met_codes) <- acronyms

    # Loop over runs to get results
    for (crop in crops) {
      scenarios <- run[[crop]]

      rPELMO_PECgw[[psm]][[crop]] <- matrix(nrow = length(scenarios), ncol = length(acronyms),
                                         dimnames = list(scenarios, acronyms))

      for (scenario in scenarios) {
        run_dir <- file.path(PELMO_base, "FOCUS", PELMO_path(psm, crop, scenario))

        psm_file <- file.path(run_dir, paste0(psm, ".psm"))
        location_code <- FOCUS_PELMO_location_codes[scenario]
        interval <- get_interval(psm_file, location_code)

        # Get percolate for each period
        wasser <- readLines(file.path(run_dir, "WASSER.PLM"))
        percolate_lines <- grep("RECHARGE BELOW ROOT ZONE", wasser, value = TRUE)
        percolate <- as.numeric(substr(percolate_lines, 40, 46))
        percolate_period <- sum_periods(percolate, interval)

        # Set up results that should match period.plm generated by the PELMO GUI
        results_rPELMO <- list()
        for (acronym in acronyms) {
          results_rPELMO[[acronym]] <- list()
          periods <- data.frame(
            period = 1:20,
            flux = NA,
            percolate = 10 * percolate_period,
            conc = NA)

          if (is.na(met_codes[acronym])) {
            chem_file <- file.path(run_dir, "CHEM.PLM")
          } else {
            chem_file <- file.path(run_dir, paste0("CHEM_", met_codes[acronym], ".PLM"))
          }

          annual_flux <- get_flux(chem_file)
          periods$flux <- sum_periods(annual_flux, interval) * 1000

          periods$conc <- 100 * periods$flux / periods$percolate

          results_rPELMO[[acronym]]$periods <- periods
          PECgw <- focus_80th(periods$conc)

          results_rPELMO[[acronym]]$focus <- PECgw

          rPELMO_PECgw[[psm]][[crop]][scenario, acronym] <- round(PECgw, 3)
        }
        save(results_rPELMO, file = file.path(run_dir, "period_rPELMO.rda"))
      }
    }
  }
  return(rPELMO_PECgw)
}

#' Get the application interval in years from a psm file
#'
#' @param psm_file The path to the .psm file
#' @param location_code The location code
get_interval <- function(psm_file, location_code) {
  # How many years do we calculate (26, 46 or 66)?
  psm <- readLines(psm_file, encoding = "latin1")
  number_of_apps_lines <- grep("number of application location", psm)
  absolute_apps_line <- grep(location_code,
                             psm[number_of_apps_lines])
  interval <- 1 # application every year
  if (length(absolute_apps_line) == 1) {
    apps_root <- number_of_apps_lines[absolute_apps_line]
  } else {
    apps_root <- number_of_apps_lines[1]
  }

  number_of_apps <- as.integer(substr(psm[apps_root], 1, 3))
  last_app_line <- psm[apps_root + number_of_apps]
  last_app_year <- as.integer(gsub("^.{2,3}  ..  (..)   .*", "\\1",
                                   last_app_line))
  if (last_app_year > 26) interval <- 2
  if (last_app_year > 46) interval <- 3
  return(interval)
}

#' Sum up values according to FOCUS periods
#'
#' @param annual The annual flux as obtained by \code{get_flux}
#' @param interval The interval in years
sum_periods <- function(annual, interval) {
  n_years <- switch(as.character(interval),
    "1" = 26,
    "2" = 46,
    "3" = 66)
  period_start_years <- seq(from = 7, to = n_years, by = interval)
  sapply(1:20, function(x) {
    years_i <- period_start_years[x] + (0 : (interval - 1))
    sum(annual[years_i])
  })
}

#' Get the flux of a chemical out of the FOCUS layer from a CHEM*.PLM file
#'
#' @param chem_file The full path to a CHEM*.PLM file
get_flux <- function(chem_file) {
  chem <- readLines(chem_file)
  lowest_focus_comp_lines <- grep("^    .     21      ", chem, value = TRUE)
  lowest_focus_comp <- read.table(text = lowest_focus_comp_lines)
  return(lowest_focus_comp$V9)
}

#' Calculate the 80th percentile according to FOCUS guidance
#'
#' This is nowadays defined as the mean of the 16th and the 17th
#' highest value. Previously, the 17th highest values was used (FOCUS 2014, p.
#' 18). NaN values need to be set to zero in order to reproduce the
#' values obtained by PELMO.
#' @param c_period A numeric vector of values to calculate the percentile from
#' @param old Should the old calculation method be used (the 17th highest value)?
#' @export
focus_80th <- function(c_period, old = FALSE) {
  c_period <- ifelse(is.na(c_period), 0, c_period)
   c_period_sorted <- sort(c_period)
  if (old) return(c_period_sorted[17])
  else return(mean(c_period_sorted[c(16, 17)]))
}
